tools/gen-unicode.py

```diff
@@ -186,25 +228,57 @@ def split_list(category_list):
     return interval_sps, interval_lengths, chars
 
 
-def generate_ranges(script_args):
-    categorizer = UnicodeCategorizer()
-    letters, non_letters, separators = categorizer.read_categories(script_args.unicode_data)
+def generate_ranges(script_args, plane_type):
+    if plane_type == "supplementary":
+        c_source = UniCodeSource(RANGES_SUP_C_SOURCE)
+        suffix = "_sup"
+        limit = 0xffff
+    else:
+        c_source = UniCodeSource(RANGES_C_SOURCE)
+        suffix = ""
+        limit = 0xff
 
-    letter_tables = split_list(list(group_ranges(letters)))
-    non_letter_tables = split_list(list(group_ranges(non_letters)))
-    separator_tables = split_list(list(group_ranges(separators)))
+    categorizer = UnicodeCategorizer()
+    categorizer.read_prop_list(script_args.prop_list)
+    letters, non_letters, separators = categorizer.read_categories(script_args.unicode_data, plane_type)
 
-    c_source = UniCodeSource(RANGES_C_SOURCE)
+    letter_tables = split_list(list(group_ranges(letters)), limit)
+    non_letter_tables = split_list(list(group_ranges(non_letters)), limit)
+    separator_tables = split_list(list(group_ranges(separators)), limit)
 
     header_completion = ["/* This file is automatically generated by the %s script" % os.path.basename(__file__),
                          " * from %s. Do not edit! */" % os.path.basename(script_args.unicode_data),
                          ""]
 
+    if plane_type == "supplementary":
+        header_completion.extend(["#if ENABLED (JERRY_ESNEXT)",
+                                  ""])
+
     c_source.complete_header("\n".join(header_completion))
 
+    array_types = {"basic": ("uint16_t",
+                             "uint8_t",
+                             "uint16_t",
+                             "uint16_t",
+                             "uint8_t",
+                             "uint16_t",
+                             "uint16_t",
+                             "uint8_t",
+                             "uint16_t"),
+                   "supplementary": ("uint32_t",
+                                     "uint16_t",
+                                     "uint32_t",
+                                     "uint32_t",
+                                     "uint16_t",
+                                     "uint32_t",
+                                     "uint32_t",
+                                     "uint16_t",
+                                     "uint32_t")}
+
     c_source.add_table(letter_tables[0],
                        "unicode_letter_interval_sps",
-                       "uint16_t",
+                       suffix,
+                       array_types[plane_type][0],
```

Comment:
This code repeated several times. Cannot we build these tables with a for loop with a list of tuples?